### PR TITLE
Update argument parsing and help/usage consistency

### PR DESCRIPTION
**Pull request by taz**
As some of the arguments required variables in `ghe-backup-config`to be set, or functions defined in that file, I ended up moving argument parsing ahead of the line that includes the config. For anything that's shared or needs to be handled inside `ghe-backup-config`, I've set flags to trigger the appropriate actions (ie for printing the --help or --version arguments).

### Changes ###

* All arguments (including -v, --version, -h and --help) can now be specified in any order
* Added consistent argument handling to `ghe-host-check`, `ghe-backup` and `ghe-restore`
* Consistency of usage text and "help" arguments improved by incorporating @cs-shadow's
changes from #297
* Added new tests to ensure --version, --help and -h are honoured in `ghe-backup`, `ghe-restore` and `ghe-host-check`

/cc @github/backup-utils for review
/cc @cs-shadow for :eyes: on the usage/help improvements

### CODE REVIEW

File: bin/ghe-backup [hunk 1]
```diff
@@ -1,12 +1,40 @@
...
 #/ Take snapshots of all GitHub Enterprise data, including Git repository data,
 #/ the MySQL database, instance settings, GitHub Pages data, etc.
 #/
-#/ With -v, enable verbose output and show more information about what's being
-#/ transferred.
+#/ Options:
+#/   -v                Enable verbose output.
+#/   --version         Display version information.
+#/   -h                Show this message.
+
```
Comment (lildude) [inline bin/ghe-backup]: Can you please change this and other locations to use the format we use for the usage in the commands included in GitHub Enterprise. The general format we use is like this:

```
#!/bin/bash
#/ Usage: ghe-maintenance [-hqsu]
#/
#/ Set or unset the maintenance mode on GitHub Enterprise
#/
#/ OPTIONS:
#/   -h | --help      Show this message
#/   -q | --query     Query current mode
#/   -s | --set       Set the maintenance mode
#/   -u | --unset     Unset the maintenance mode
#/
```

Specific things to note:

- grouping of short options in the initial `Usage` line
- OPTIONS is capitalised
- Short & long options are shown in the OPTIONS section, where available.

**Comment by taz** [comment]
I found a small argument parsing bug in `ghe-restore` that I'm looking into (it's ignoring arguments added after the snapshot id)... I'll push some new code to fixing that, and the usage text format tomorrow.

**Comment by taz** [comment]
@lildude I've cleaned up and re-arranged the usage text a bit as suggested. I also added some additional longer arguments for consistency too.

File: bin/ghe-restore [hunk 2]
```diff
@@ -1,57 +1,75 @@
...
+#/ Usage: ghe-restore [-fchv] [--version] [-s <snapshot-id>] [<host>]
 #/
-#/ Options:
-#/   --version         Display version information.
-#/   -f                Don't prompt for confirmation before restoring.
-#/   -c                Restore appliance settings and license in addition to
+#/ Restores a GitHub instance from local backup snapshots.
+#/
+#/ Note that the GitHub Enterprise host must be reachable and your SSH key must
+#/ be setup asdescribed in the following help article:
```
Comment (lildude) [inline bin/ghe-restore]: Need a space in "asdescribed"

File: bin/ghe-backup [hunk 3]
```diff
@@ -1,12 +1,42 @@
...
+#/
 #/ Take snapshots of all GitHub Enterprise data, including Git repository data,
 #/ the MySQL database, instance settings, GitHub Pages data, etc.
 #/
-#/ With -v, enable verbose output and show more information about what's being
-#/ transferred.
+#/ OPTIONS:
+#/   -v | --verbose    Enable verbose output
+#/   -h | --help       Show this message
+#/        --version    Display version information
```
Comment (lildude) [inline bin/ghe-backup]: Nit: need full stops at the end of each of these to be consistent with elsewhere.

File: test/test-ghe-backup.sh [hunk 4]
```diff
@@ -524,3 +524,26 @@ begin_test "ghe-backup with no leaked keys"
...
+
+)
+end_test
+
+begin_test "ghe-backup honours --help and -h flags"
+(
+  set -e
+
+  arg_help=`ghe-backup --help | grep -o 'Usage: ghe-backup'`
+  arg_h=`ghe-backup -h | grep -o 'Usage: ghe-backup'`
```
Comment (lildude) [inline test/test-ghe-backup.sh]: We prefer `$()` over backticks. 

Reminder to myself: add this to the styleguide.

File: test/test-ghe-host-check.sh [hunk 5]
```diff
@@ -24,3 +24,26 @@ begin_test "ghe-host-check with host arg"
...
+
+)
+end_test
+
+begin_test "ghe-host-check honours --help and -h flags"
+(
+  set -e
+
+  arg_help=`ghe-host-check --help | grep -o 'Usage: ghe-host-check'`
+  arg_h=`ghe-host-check -h | grep -o 'Usage: ghe-host-check'`
```
Comment (lildude) [inline test/test-ghe-host-check.sh]: As above, we prefer `$()` over backticks.

File: test/test-ghe-restore.sh [hunk 6]
```diff
@@ -653,6 +653,28 @@ begin_test "ghe-restore fails when restore to an active HA pair"
...
+
+)
+end_test
+
+begin_test "ghe-restore honours --help and -h flags"
+(
+  set -e
+
+  arg_help=`ghe-restore --help | grep -o 'Usage: ghe-restore'`
+  arg_h=`ghe-restore -h | grep -o 'Usage: ghe-restore'`
```
Comment (lildude) [inline test/test-ghe-restore.sh]: As above, we prefer `$()` over backticks.

File: test/test-ghe-backup.sh (hunk 4 above)
Comment (taz) [reply test/test-ghe-backup.sh]: Good to know! Fixed.

**Pull request merged**